crates/sui-indexer/src/processors/object_balance_processor.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_json_rpc_types::{CheckpointId, SuiTransactionBlockEffects, SuiTransactionBlockEffectsAPI};
+use tracing::info;
+
+use crate::errors::IndexerError;
+use crate::models::object_balances::ObjectBalance;
+use crate::store::IndexerStore;
+
+pub const OBJECT_BALANCES_WATERMARK: &str = "object-balances";
+
+pub struct ObjectBalanceProcessor<S> {
+    pub store: S,
+}
+
+impl<S> ObjectBalanceProcessor<S>
+where
+    S: IndexerStore + Sync + Send + 'static,
+{
+    pub fn new(store: S) -> Self {
+        Self { store }
+    }
+
+    pub async fn start(&self) -> Result<(), IndexerError> {
+        info!("Indexer ObjectBalance async processor started...");
+        let watermark = self.store.get_watermark(OBJECT_BALANCES_WATERMARK).await?;
+        let mut cursor = watermark
+            .map(|w| w.checkpoint.map(|c| c.saturating_add(1)).unwrap_or(0))
+            .unwrap_or(0);
+        loop {
+            let latest_checkpoint = self.store.get_latest_checkpoint_sequence_number().await?;
+
+            if latest_checkpoint < cursor {
+                tokio::time::sleep(tokio::time::Duration::from_secs(3)).await;
```

Comment:
sleep for 0.5 or 1 second as it takes less than 3 seconds to generate a new checkpoint?